installer/aws_cluster.go

```diff
@@ -0,0 +1,609 @@
+package installer
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/awslabs/aws-sdk-go/aws"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/awslabs/aws-sdk-go/gen/cloudformation"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/awslabs/aws-sdk-go/gen/ec2"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/awslabs/aws-sdk-go/gen/route53"
+	"github.com/flynn/flynn/pkg/awsutil"
+	"github.com/flynn/flynn/pkg/sshkeygen"
+	"github.com/flynn/flynn/util/release/types"
+)
+
+var DisallowedEC2InstanceTypes = []string{"t1.micro", "t2.micro", "t2.small", "m1.small"}
+var DefaultInstanceType = "m3.medium"
+var StackNotFoundError = errors.New("Stack does not exist")
+
+func (c *AWSCluster) Base() *BaseCluster {
+	return c.base
+}
+
+func (c *AWSCluster) wrapRequest(runRequest func() error) error {
+	const rateExceededErrStr = "Rate exceeded"
+	const maxBackoff = 10 * time.Second
+	backoff := 1 * time.Second
+	timeout := time.After(35 * time.Second)
+	for {
+		err := runRequest()
+		if err != nil && err.Error() == rateExceededErrStr {
+			select {
+			case <-time.After(backoff):
+				backoff = backoff * 2
+				if backoff > maxBackoff {
+					backoff = maxBackoff
+				}
+				continue
+			case <-timeout:
+			}
+		}
+		return err
+	}
+}
+
+func (c *AWSCluster) saveField(field string, value interface{}) error {
+	c.base.installer.dbMtx.Lock()
+	defer c.base.installer.dbMtx.Unlock()
+
+	tx, err := c.base.installer.db.Begin()
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(fmt.Sprintf(`
+  UPDATE aws_clusters SET %s = $2 WHERE ClusterID == $1
+  `, field), c.ClusterID, value)
```

Comment:
ql gives an error for any mutation attempted outside a transaction, would it be better to make that part of the exec string?